File: luda/editor/client/src/cameraAngleEditor/wysiwygEditor/Resizer.ts
import {
  ColorUtil,
  RenderingTree,
  Translate,
  Vector,
  PathDrawCommand,
  RenderingData,
  Cursor,
  engine,
  Rect,
  AfterDraw,
  Convert,
  Render,
} from "namui";
import { LtrbRect } from "namui/lib/type";
import { CameraAngleEditorState } from "../type";

export const Resizer: Render<
  CameraAngleEditorState,
  {
    widthHeightRatio: number;
  }
> = (state, props) => {
  return [
    Rect({
      ...state.cameraAngle.sourceRect,
      style: {
        stroke: {
          color: ColorUtil.Grayscale01(0.2),
          width: 1,
        },
      },
    }),
    Translate(
      {
        x: state.cameraAngle.sourceRect.x,
        y: state.cameraAngle.sourceRect.y,
      },
      [
        renderImageSizeHandles(
          {
            widthHeightRatio: props.widthHeightRatio,
          },
          state,
        ),
      ],
    ),
  ];
};

function renderImageSizeHandles(
  props: {
    widthHeightRatio: number;
  },
  state: CameraAngleEditorState,
): RenderingTree {
  const { sourceRect } = state.cameraAngle;
  const handleRadius = 5;
  const handles: {
    id:
      | "resize-top-left"
      | "resize-top"
      | "resize-top-right"
      | "resize-left"
      | "resize-right"
      | "resize-bottom-left"
      | "resize-bottom"
      | "resize-bottom-right";
    position: Vector;
    cursor: Cursor;
  }[] = [
    {
      id: "resize-top-left",
      position: new Vector(0, 0),
      cursor: Cursor.leftTopRightBottomResize,
    },
    {
      id: "resize-top",
      position: new Vector(sourceRect.width / 2, 0),
      cursor: Cursor.topBottomResize,
    },
    {
      id: "resize-top-right",
      position: new Vector(sourceRect.width, 0),
      cursor: Cursor.rightTopLeftBottomResize,
    },
    {
      id: "resize-left",
      position: new Vector(0, sourceRect.height / 2),
      cursor: Cursor.leftRightResize,
    },
    {
      id: "resize-right",
      position: new Vector(sourceRect.width, sourceRect.height / 2),
      cursor: Cursor.leftRightResize,
    },
    {
      id: "resize-bottom-left",
      position: new Vector(0, sourceRect.height),
      cursor: Cursor.rightTopLeftBottomResize,
    },
    {
      id: "resize-bottom",
      position: new Vector(sourceRect.width / 2, sourceRect.height),
      cursor: Cursor.topBottomResize,
    },
    {
      id: "resize-bottom-right",
      position: new Vector(sourceRect.width, sourceRect.height),
      cursor: Cursor.leftTopRightBottomResize,
    },
  ];

  return [
    ...handles.map((handle) => {
      const left = handle.position.x - handleRadius;
      const top = handle.position.y - handleRadius;
      const right = handle.position.x + handleRadius;
      const bottom = handle.position.y + handleRadius;

      const path = new CanvasKit.Path();
      path.addOval([left, top, right, bottom]);

      const fillPaint = new CanvasKit.Paint();
      fillPaint.setStyle(CanvasKit.PaintStyle.Fill);
      fillPaint.setColor(ColorUtil.White);

      const strokePaint = new CanvasKit.Paint();
      strokePaint.setStyle(CanvasKit.PaintStyle.Stroke);
      strokePaint.setColor(ColorUtil.Grayscale01(0.5));
      strokePaint.setStrokeWidth(2);
      strokePaint.setAntiAlias(true);

      return RenderingData({
        drawCalls: [
          {
            commands: [
              PathDrawCommand({
                path,
                paint: fillPaint,
              }),
              PathDrawCommand({
                path,
                paint: strokePaint,
              }),
            ],
          },
        ],
        onMouseIn() {
          engine.mousePointer.setCursor(handle.cursor);
        },
        onMouseDown(event) {
          state.wysiwygEditor.dragging = {
            targetId: handle.id,
            lastMousePosition: Vector.from(event),
          };
        },
      });
    }),
    AfterDraw(({ translated }) => {
      const container = new Vector(
        translated.x - sourceRect.x,
        translated.y - sourceRect.y,
      );
      engine.mouseEvent.onMouseMove((event) => {
        const { dragging } = state.wysiwygEditor;
        if (!dragging || !dragging.targetId.startsWith("resize-")) {
          return;
        }
        const { targetId } = dragging;
        const mouseVector = Vector.from(event);
        const diff = mouseVector.sub(dragging.lastMousePosition);

        const mouseLocalVector = mouseVector.sub(container);
        const handleId = targetId.substring("resize-".length);
        const isDiagonal = targetId.includes("-");

        if (isDiagonal) {
          const centerPoint = new Vector(
            sourceRect.x + sourceRect.width / 2,
            sourceRect.y + sourceRect.height / 2,
          );
          const endPoint = new Vector(
            handleId.includes("left")
              ? sourceRect.x
              : sourceRect.x + sourceRect.width,
            handleId.includes("top")
              ? sourceRect.y
              : sourceRect.y + sourceRect.height,
          );

          const diagonalVector = endPoint.sub(centerPoint);
          const centerToMouseLocalVector = mouseLocalVector.sub(centerPoint);

          const isCcw = diagonalVector.cross(centerToMouseLocalVector) > 0;
          function getMain(): keyof LtrbRect {
            switch (handleId) {
              case "top-left":
                return isCcw ? "top" : "left";
              case "top-right":
                return isCcw ? "right" : "top";
              case "bottom-right":
                return isCcw ? "bottom" : "right";
              case "bottom-left":
                return isCcw ? "left" : "bottom";
              default:
                throw new Error("unreachable");
            }
          }
          const main = getMain();
          const sub = handleId
            .replace(main, "")
            .replace("-", "") as keyof LtrbRect;

          const nextLtrbRect = resizeRect({
            rect: Convert.xywhToLtrb(sourceRect),
            main,
            sub,
            mouseLocalVector,
            widthHeightRatio: props.widthHeightRatio,
          });
          const nextRect = Convert.ltrbToXywh(nextLtrbRect);
          sourceRect.x = nextRect.x;
          sourceRect.y = nextRect.y;
          sourceRect.width = nextRect.width;
          sourceRect.height = nextRect.height;
        } else {
          if (handleId === "top" || handleId === "bottom") {
            let deltaWidth: number;

            if (handleId === "top") {
              sourceRect.y += diff.y;
              sourceRect.height -= diff.y;
              deltaWidth = -diff.y * props.widthHeightRatio;
            } else {
              sourceRect.height += diff.y;
              deltaWidth = diff.y * props.widthHeightRatio;
            }

            sourceRect.x -= deltaWidth / 2;
            sourceRect.width += deltaWidth;
          } else {
            let deltaHeight: number;

            if (handleId === "left") {
              sourceRect.x += diff.x;
              sourceRect.width -= diff.x;
              deltaHeight = -diff.x / props.widthHeightRatio;
            } else {
              sourceRect.width += diff.x;
              deltaHeight = diff.x / props.widthHeightRatio;
            }

            sourceRect.y -= deltaHeight / 2;
            sourceRect.height += deltaHeight;
          }
        }
        dragging.lastMousePosition = mouseVector;
      });
    }),
  ];
}

function resizeRect({
  rect,
  main,
  sub,
  mouseLocalVector,
  widthHeightRatio,
}: {
  rect: LtrbRect;
  main: keyof LtrbRect;
  sub: keyof LtrbRect;
  mouseLocalVector: Vector;
  widthHeightRatio: number;
}): LtrbRect {
  const nextRect: LtrbRect = {
    ...rect,
  };

  const isMainHorizontal = main === "left" || main === "right";

  nextRect[main] = isMainHorizontal ? mouseLocalVector.x : mouseLocalVector.y;
  const deltaMain = nextRect[main] - rect[main];

  const shouldDeltaReverse =
    (main === "right" && sub === "top") ||
    (main === "top" && sub === "right") ||
    (main === "left" && sub === "bottom") ||
    (main === "bottom" && sub === "left");

  nextRect[sub] +=
    deltaMain *
    (isMainHorizontal ? 1 / widthHeightRatio : widthHeightRatio) *
    (shouldDeltaReverse ? -1 : 1);

  return nextRect;
}

File: luda-editor/luda-editor-client/src/app/editor/clip_editor/camera_clip_editor_legacy/wysiwyg_editor/resizer.rs
use crate::app::editor::clip_editor::camera_clip_editor::wysiwyg_editor::WysiwygEvent;
use namui::prelude::*;

pub struct Resizer {
    id: String,
}

impl Resizer {
    pub fn new(id: &str) -> Self {
        Self { id: id.to_string() }
    }
}

pub struct ResizerProps<'a> {
    pub source_rect: &'a XywhRect<f32>,
    pub container_size: &'a Wh<f32>,
}

impl Resizer {
    pub fn update(&mut self, event: &dyn std::any::Any) {}

    pub fn render(&self, props: &ResizerProps) -> RenderingTree {
        render![
            rect(RectParam {
                x: props.source_rect.x,
                y: props.source_rect.y,
                width: props.source_rect.width,
                height: props.source_rect.height,
                style: RectStyle {
                    stroke: Some(RectStroke {
                        color: Color::grayscale_f01(0.2),
                        width: 1.0,
                        border_position: BorderPosition::Inside,
                    }),
                    ..Default::default()
                },
            }),
            self.render_resize_handles(props.source_rect, props.container_size),
        ]
    }

    fn render_resize_handles(
        &self,
        source_rect: &XywhRect<f32>,
        container_size: &Wh<f32>,
    ) -> RenderingTree {
        const HANDLE_RADIUS: f32 = 5.0;
        RenderingTree::Children(
            get_handles(&source_rect)
                .iter()
                .map(|handle| {
                    let path = namui::PathBuilder::new().add_oval(&LtrbRect {
                        left: handle.xy.x - HANDLE_RADIUS,
                        top: handle.xy.y - HANDLE_RADIUS,
                        right: handle.xy.x + HANDLE_RADIUS,
                        bottom: handle.xy.y + HANDLE_RADIUS,
                    });

                    let fill_paint = namui::PaintBuilder::new()
                        .set_style(namui::PaintStyle::Fill)
                        .set_color(Color::WHITE);

                    let stroke_paint = namui::PaintBuilder::new()
                        .set_style(namui::PaintStyle::Stroke)
                        .set_color(Color::grayscale_f01(0.5))
                        .set_stroke_width(2.0)
                        .set_anti_alias(true);

                    let resizer_id = self.id.clone();

                    render![
                        namui::path(path.clone(), fill_paint),
                        namui::path(path, stroke_paint),
                    ]
                    .with_mouse_cursor(handle.cursor())
                    .attach_event(move |builder| {
                        let resizer_id = resizer_id.clone();
                        let handle = handle.clone();
                        let container_size = container_size.clone();
                        let source_rect = source_rect.clone();
                        builder.on_mouse_down(move |mouse_event| {
                            namui::event::send(WysiwygEvent::ResizerHandleMouseDownEvent {
                                target_id: resizer_id.clone(),
                                handle,
                                center_xy: source_rect.center(),
                                mouse_xy: mouse_event.global_xy,
                                container_size,
                                image_size_ratio: Wh {
                                    width: source_rect.width,
                                    height: source_rect.height,
                                },
                            })
                        })
                    })
                })
                .collect::<Vec<RenderingTree>>(),
        )
    }
}

fn get_opposite_handle(handle: &ResizerHandle, source_rect: &XywhRect<f32>) -> ResizerHandle {
    let center_xy = source_rect.center();
    ResizerHandle {
        handle_direction: handle.handle_direction.opposite(),
        xy: 2.0 * center_xy - handle.xy,
    }
}

#[derive(Debug, Clone, Copy)]
pub enum ResizerHandleDirection {
    TopLeft,
    TopRight,
    BottomRight,
    BottomLeft,
    Top,
    Right,
    Bottom,
    Left,
}
#[derive(Debug, Clone, Copy)]
pub struct ResizerHandle {
    pub handle_direction: ResizerHandleDirection,
    pub xy: Xy<f32>,
}
impl ResizerHandle {
    pub fn cursor(&self) -> namui::MouseCursor {
        self.handle_direction.cursor()
    }
}
fn get_handles(source_rect: &XywhRect<f32>) -> Vec<ResizerHandle> {
    let top_left = ResizerHandle {
        handle_direction: ResizerHandleDirection::TopLeft,
        xy: Xy {
            x: source_rect.x,
            y: source_rect.y,
        },
    };

    let top = ResizerHandle {
        handle_direction: ResizerHandleDirection::Top,
        xy: Xy {
            x: source_rect.x + source_rect.width / 2.0,
            y: source_rect.y,
        },
    };
    let top_right = ResizerHandle {
        handle_direction: ResizerHandleDirection::TopRight,
        xy: Xy {
            x: source_rect.x + source_rect.width,
            y: source_rect.y,
        },
    };
    let left = ResizerHandle {
        handle_direction: ResizerHandleDirection::Left,
        xy: Xy {
            x: source_rect.x,
            y: source_rect.y + source_rect.height / 2.0,
        },
    };
    vec![
        top_left,
        top,
        top_right,
        left,
        get_opposite_handle(&left, &source_rect),
        get_opposite_handle(&top_right, &source_rect),
        get_opposite_handle(&top, &source_rect),
        get_opposite_handle(&top_left, &source_rect),
    ]
}
impl ResizerHandleDirection {
    pub(crate) fn opposite(&self) -> Self {
        match self {
            ResizerHandleDirection::TopLeft => ResizerHandleDirection::BottomRight,
            ResizerHandleDirection::TopRight => ResizerHandleDirection::BottomLeft,
            ResizerHandleDirection::BottomLeft => ResizerHandleDirection::TopRight,
            ResizerHandleDirection::BottomRight => ResizerHandleDirection::TopLeft,
            ResizerHandleDirection::Top => ResizerHandleDirection::Bottom,
            ResizerHandleDirection::Bottom => ResizerHandleDirection::Top,
            ResizerHandleDirection::Left => ResizerHandleDirection::Right,
            ResizerHandleDirection::Right => ResizerHandleDirection::Left,
        }
    }
    pub(crate) fn cursor(&self) -> MouseCursor {
        match self {
            ResizerHandleDirection::TopLeft => MouseCursor::LeftTopRightBottomResize,
            ResizerHandleDirection::Top => MouseCursor::TopBottomResize,
            ResizerHandleDirection::TopRight => MouseCursor::RightTopLeftBottomResize,
            ResizerHandleDirection::Left => MouseCursor::LeftRightResize,
            ResizerHandleDirection::Right => MouseCursor::LeftRightResize,
            ResizerHandleDirection::BottomLeft => MouseCursor::RightTopLeftBottomResize,
            ResizerHandleDirection::Bottom => MouseCursor::TopBottomResize,
            ResizerHandleDirection::BottomRight => MouseCursor::LeftTopRightBottomResize,
        }
    }
}
